### PR TITLE
qa/tasks/cephadm: stop ceph-{fsid}.target on shutdown

### PR DESCRIPTION
**Pull request by liewegas**
Daemons can be deployed by cephadm directly that are unknown to us
(that aren't listed in our roles dict).  Stop them too.  Among other
things, this can avoid problems while gzipping their logs.

Signed-off-by: Sage Weil <sage@newdream.net>

### CODE REVIEW

File: qa/tasks/cephadm.py [hunk 1]
```diff
@@ -477,7 +475,17 @@ def ceph_bootstrap(ctx, config):
...
                 raise
 
+        # stop the catch-all, too, so that we also stop daemons/roles we
+        # don't know about  Note that doesn't block until they are all stopped, so
+        # wait a moment
+        log.info(f'Stopping ceph-{fsid}.target...')
+        ctx.cluster.run(args=['sudo', 'systemctl', 'stop', f'ceph-{fsid}.target'])
+
+        log.info('Waiting a few moments for containers to stop...')
+        time.sleep(10)
```
Comment (jdurgin) [inline qa/tasks/cephadm.py:485]: sleeps like this seem prone to racing with e.g. log gathering, I suppose we can see how much it comes up in practice though

**Review by jdurgin — APPROVED** [review]
concerned about the sleep hiding races, otherwise looks good

**Comment by liewegas** [comment]
> concerned about the sleep hiding races, otherwise looks good

'systemctl stop something.target' does not wait for the constituent services to stop :(, and I can't find any other systemctl magic to do so.  I think to do this properly we need enumerate the running services and stop them individually?  Having a hard time believing systemd can't do this for us...

**Comment by liewegas** [comment]
- could add a ``--keep-logs`` option to ``cephadm rm-cluster`` and run that first, then gather logs, then rm-cluster again to blow away everything
- could add a ``stop-cluster`` command to cephadm that just stops all of the units.  

The annoying thing with both of those options is that we need to either backport quickly or teach cephadm.py how to behave (sleep?) if the cephadm version doesn't implement those commands.

**Comment by jdurgin** [comment]
Looks like our unit files aren't following the systemd interface - ```ExecStop=-{container_path} stop ceph-{fsid}-%i``` - according to systemd docs the script run by ExecStop should stop the process(es) synchronously.

https://www.freedesktop.org/software/systemd/man/systemd.service.html
```Note that it is usually not sufficient to specify a command for this setting that only asks the service to terminate (for example, by sending some form of termination signal to it), but does not wait for it to do so. Since the remaining processes of the services are killed according to KillMode= and KillSignal= or RestartKillSignal= as described above immediately after the command exited, this may not result in a clean stop. The specified command should hence be a synchronous operation, not an asynchronous one.```

I also see '{podman|docker} stop' sends SIGTERM, then SIGKILL after 10s. That's probably not what we want either...

**Comment by jdurgin** [comment]
> > concerned about the sleep hiding races, otherwise looks good
> 
> 'systemctl stop something.target' does not wait for the constituent services to stop :(, and I can't find any other systemctl magic to do so. I think to do this properly we need enumerate the running services and stop them individually? Having a hard time believing systemd can't do this for us...

It looks like using PartOf instead of WantedBy in the unit files might do the trick - affecting starting/stopping in only one direction.

**Comment by liewegas** [comment]
```
2021-03-23T22:58:10.380 INFO:tasks.cephadm:Stopping ceph-c99f8f4a-8c2a-11eb-90a7-001a4aab830c.target...
2021-03-23T22:58:10.380 DEBUG:teuthology.orchestra.run.smithi052:> sudo systemctl stop ceph-c99f8f4a-8c2a-11eb-90a7-001a4aab830c.target
2021-03-23T22:58:10.401 DEBUG:teuthology.orchestra.run.smithi105:> sudo systemctl stop ceph-c99f8f4a-8c2a-11eb-90a7-001a4aab830c.target
2021-03-23T22:58:10.421 INFO:teuthology.orchestra.run.smithi105.stderr:Failed to stop ceph-c99f8f4a-8c2a-11eb-90a7-001a4aab830c.target: Unit ceph-c99f8f4a-8c2a-11eb-90a7-001a4aab830c.target not loaded.
2021-03-23T22:58:10.423 DEBUG:teuthology.orchestra.run:got remote process result: 5
```
/a/sage-2021-03-23_22:31:12-rados:cephadm-wip-sage-testing-2021-03-23-1606-distro-basic-smithi/5990953

**Comment by liewegas** [comment]
replaced by a commit in #40888

**Pull request closed without merging**